pkg/endpoint/redirect_test.go

```diff
@@ -478,6 +482,315 @@ func (s *RedirectSuite) TestRedirectWithDeny(c *check.C) {
+		mapKeyAllowAllE: {
+			DerivedFromRules: labels.LabelArrayList{AllowAnyEgressLabels},
+		},
+		mapKeyFooL7: {
+			ProxyPort:        crd2Port,
+			Listener:         "/cec2/listener2",
+			DerivedFromRules: labels.LabelArrayList{lblsL4AllowListener1, lblsL4AllowPort80, lblsL4L7AllowListener2Priority1},
+		},
+		mapKeyAllL7: {
+			DerivedFromRules: labels.LabelArrayList{lblsL4AllowListener1, lblsL4AllowPort80, lblsL4L7AllowListener2Priority1},
+		},
+	})
+	if !ep.desiredPolicy.GetPolicyMap().Equals(expected2) {
+		c.Fatal("desired policy map does not equal expected map:\n",
+			ep.desiredPolicy.GetPolicyMap().Diff(c.T, expected2))
+	}
+
+	// Keep only desired redirects
+	ep.removeOldRedirects(desiredRedirects, realizedRedirects, cmp)
+
+	// Check that the redirect is still realized
+	c.Assert(len(desiredRedirects), check.Equals, 2)
+	c.Assert(ep.desiredPolicy.GetPolicyMap().Len(), check.Equals, 3)
+
+	// Pretend that something failed and revert the changes
+	revertFunc()
+
+	// Check that the state before addRedirects is restored
+	if !ep.desiredPolicy.GetPolicyMap().Equals(expected) {
+		c.Fatal("desired policy map does not equal expected map:\n",
+			ep.desiredPolicy.GetPolicyMap().Diff(c.T, expected))
+	}
+	c.Assert(ep.desiredPolicy.GetPolicyMap().Len(), check.Equals, 2)
+}
+
+func (s *RedirectSuite) TestRedirectWithEqualPriority(c *check.C) {
+	// Setup dependencies for endpoint.
+	kvstore.SetupDummy(c, "etcd")
+
+	oldPolicyEnable := policy.GetPolicyEnabled()
+	defer policy.SetPolicyEnabled(oldPolicyEnable)
+	policy.SetPolicyEnabled(option.DefaultEnforcement)
+
+	idAllocatorOwner := &DummyIdentityAllocatorOwner{}
+
+	mgr := cache.NewCachingIdentityAllocator(idAllocatorOwner)
+	<-mgr.InitIdentityAllocator(nil)
+	defer mgr.Close()
+
+	identityCache := cache.IdentityCache{
+		identity.NumericIdentity(identityFoo): labelsFoo,
+		identity.NumericIdentity(identityBar): labelsBar,
+	}
+
+	do := &DummyOwner{
+		repo: policy.NewPolicyRepository(mgr, identityCache, nil, nil),
+	}
+	do.repo.GetSelectorCache().SetLocalIdentityNotifier(testidentity.NewDummyIdentityNotifier())
+	identitymanager.Subscribe(do.repo)
+	defer identitymanager.RemoveAll()
+
+	crd1Port := uint16(19000)
+	crd2Port := uint16(19001)
+
+	rsp := &RedirectSuiteProxy{
+		parserProxyPortMap: map[string]uint16{
+			"crd/cec1/listener1": crd1Port,
+			"crd/cec2/listener2": crd2Port,
+		},
+		redirectPortUserMap: make(map[uint16][]string),
+	}
+
+	ep := NewEndpointWithState(do, do, testipcache.NewMockIPCache(), rsp, mgr, 12345, StateRegenerating)
+
+	epIdentity, _, err := mgr.AllocateIdentity(context.Background(), labelsBar.Labels(),
+		true, identity.NumericIdentity(identityBar))
+	c.Assert(err, check.IsNil)
+	ep.SetIdentity(epIdentity, true)
+
+	rules := api.Rules{
+		ruleL4L7AllowListener1Priority1.WithEndpointSelector(selectBar_),
+		ruleL4AllowPort80.WithEndpointSelector(selectBar_),
+		ruleL4L7AllowListener2Priority1.WithEndpointSelector(selectBar_),
+	}
+	do.repo.AddList(rules)
```

Comment:
As I understand, this test was written in response to questions about this case:

- Traffic to app=foo on port 80/TCP should use listener A with prio 1
- Traffic to class=empire on port 80/TCP should use listener B with prio 1
- Each of these rules to select the same identity if that identity includes both app=foo,class=empire

(cf. https://github.com/cilium/cilium/pull/28555#discussion_r1445504871)

These rules seem to cover the "same priority" aspect of testing, using the same selector. Does this test also cover the "disjoint selector" aspect? The reason I ask is that I can understand and saw the code that handles having a single selector which gets merged with another selector due to the selector+port collision, but I didn't understand whether it covers the case where the selectors end up distinct inside `l4.PerSelectorPolicies`.

Reply:
`WithEndpointSelector()` here specifies which endpoints the rule applies to:
```
type Rule struct {
	// EndpointSelector selects all endpoints which should be subject to
	// this rule. EndpointSelector and NodeSelector cannot be both empty and
	// are mutually exclusive.
	//
	// +kubebuilder:validation:OneOf
	EndpointSelector EndpointSelector `json:"endpointSelector,omitempty"`
```

The rules themselves have different `FromEndpoints` selectors as intended:
- `ruleL4L7AllowListener1Priority1`: `FromEndpoints: []api.EndpointSelector{selectFoo_}`
- `ruleL4L7AllowListener2Priority1`: `FromEndpoints: []api.EndpointSelector{selectRed_}`

Reply:
Ah right. One of these days we should come up with canonical terminology for the target policy enforcement point vs. the peer selector. Makes sense, thanks!